cli/src/main/java/org/projectnessie/tools/catalog/migration/cli/CatalogMigrationCLI.java

```diff
@@ -0,0 +1,507 @@
+
+    Configuration targetCatalogConf = new Configuration();
+    if (targetHadoopConf != null && !targetHadoopConf.isEmpty()) {
+      targetHadoopConf.forEach(targetCatalogConf::set);
+    }
+    Catalog targetCatalog =
+        CatalogUtil.loadCatalog(
+            Objects.requireNonNull(catalogImpl(targetCatalogType, targetCustomCatalogImpl)),
+            targetCatalogType.name(),
+            targetCatalogProperties,
+            targetCatalogConf);
+    printWriter.println(String.format("%nConfigured target catalog: %s", targetCatalogType.name()));
+
+    List<TableIdentifier> tableIdentifiers = processIdentifiersInput(printWriter);
+
+    if (!canProceed(printWriter, sourceCatalog)) {
+      return 0;
+    }
+
+    ImmutableCatalogMigratorParams params =
+        ImmutableCatalogMigratorParams.builder()
+            .sourceCatalog(sourceCatalog)
+            .targetCatalog(targetCatalog)
+            .deleteEntriesFromSourceCatalog(deleteSourceCatalogTables)
+            .build();
+    CatalogMigrator catalogMigrator = new CatalogMigrator(params);
+
+    List<TableIdentifier> identifiers;
+    if (tableIdentifiers.isEmpty()) {
+      if (identifiersRegEx == null) {
+        printWriter.println(
+            String.format(
+                "%nUser has not specified the table identifiers."
+                    + " Selecting all the tables from all the namespaces from the source catalog."));
+      } else {
+        printWriter.println(
+            String.format(
+                "%nUser has not specified the table identifiers."
+                    + " Selecting all the tables from all the namespaces from the source catalog "
+                    + "which matches the regex pattern:"
+                    + identifiersRegEx));
+      }
+      identifiers = catalogMigrator.getMatchingTableIdentifiers(identifiersRegEx);
+    } else {
+      identifiers = tableIdentifiers;
+    }
+
+    String operation = deleteSourceCatalogTables ? "migration" : "registration";
+    printWriter.println(
+        String.format("%nIdentified %d tables for %s.", identifiers.size(), operation));
+
+    ImmutableCatalogMigrationResult.Builder resultBuilder =
+        ImmutableCatalogMigrationResult.builder();
+    if (isDryRun) {
+      CatalogMigrationResult result =
+          resultBuilder.addAllRegisteredTableIdentifiers(identifiers).build();
+      writeToFile(pathWithOutputDir(DRY_RUN_FILE), result.registeredTableIdentifiers());
+      printWriter.println("Dry run is completed.");
+      printDryRunResults(result, printWriter);
+      return 0;
+    }
+
+    printWriter.println(String.format("%nStarted %s ...", operation));
+
+    List<List<TableIdentifier>> IdentifierBatches = Lists.partition(identifiers, BATCH_SIZE);
+    AtomicInteger counter = new AtomicInteger();
+    IdentifierBatches.forEach(
+        identifierBatch -> {
+          catalogMigrator.registerTables(identifierBatch, resultBuilder);
```

Comment:
That `CatalogMigrator` should only maintain the result instance being built and eventually return the built instance - not receive a builder instance.

Reply:
In that case, each batch will return the lists (from a new builder), to avoid copying each batch's list into a bigger list, I passed the builder itself. 

I can change it back.

Reply:
CatalogMigrator should maintain the result instance. NObody talked about moving collections around.